proparse/src/main/java/org/prorefactor/proparse/antlr4/TokenList.java

```diff
@@ -50,99 +49,110 @@
+      if (nxt.getType() == NodeTypes.OBJCOLON) {
         reviewObjcolon();
-      if (nextToken.getType() == PreprocessorParser.EOF)
+      }
+      if ((nxt.getType() == NodeTypes.OBJCOLON) || (nxt.getType() == Token.EOF))
         break;
+      nxt = (ProToken) source.nextToken();
     }
-    LOGGER.trace("Exiting TokenList#build() - {} tokens", list.size());
   }
 
   private void reviewObjcolon() {
-    int colonIndex = list.size() - 1;
-    int lastIndex = colonIndex - 1;
+    ProToken objColonToken = queue.removeLast();
+    Deque<ProToken> comments = new LinkedList<>();
+    Deque<ProToken> clsName = new LinkedList<>();
 
-    // Getting type of the token just before colon (excluding comments and whitespaces)
-    int ttype = list.get(lastIndex).getType();
-    while (ttype == PreprocessorParser.WS || ttype == PreprocessorParser.COMMENT) {
-      ttype = list.get(--lastIndex).getType();
-    }
-
-    // Look for NAMEDOT pairs.
-    // Actually, it's not that easy. Something like:
-    // newsyntax.101b.deep.FindMe
-    // is perfectly valid, and because of the digit following the '.',
-    // one part of that name gets picked up as a token with text ".101b".
-    int index = lastIndex;
     boolean foundNamedot = false;
-    for (;;) {
-      if (index == 0)
-        break;
-      int currType = list.get(index).getType();
-      if (currType == PreprocessorParser.WS || currType == PreprocessorParser.COMMENT) {
-        // There can be space in front of a NAMEDOT in a table or field name.
-        // We don't want to fiddle with those here.
-        return;
+    ProToken tok = null;
+    try {
+      // Store comments and whitespaces before the colon
+      tok = queue.removeLast();
+      while ((tok.getType() == NodeTypes.WS) || (tok.getType() == NodeTypes.COMMENT)) {
+        comments.addFirst(tok);
+        tok = queue.pollLast();
       }
-      if (list.get(index - 1).getType() == PreprocessorParser.NAMEDOT) {
-        index = index - 2;
-      } else if (list.get(index).getText().charAt(0) == '.') {
-        index = index - 1;
-      } else {
-        break;
+  
+      foundNamedot = false;
+      while (true) {
+        // There can be space in front of a NAMEDOT in a table or field name. We don't want to fiddle with those here.
+        if ((tok.getType() == PreprocessorParser.WS) || (tok.getType() == PreprocessorParser.COMMENT)) {
+          break;
+        }
+  
+        // If previous is NAMEDOT, then we add both tokens
+        if ((queue.peekLast() != null) && (queue.peekLast().getType() == NodeTypes.NAMEDOT)) {
+          clsName.addFirst(tok);
+          clsName.addFirst(queue.removeLast());
+          tok = queue.removeLast();
+        } else if (tok.getText().startsWith(".")) {
+          clsName.addFirst(tok);
+          tok = queue.removeLast();
+        } else {
+          break;
+        }
+        foundNamedot = true;
       }
-      foundNamedot = true;
+    } catch (NoSuchElementException caught) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS1166)